File: src/main/scala/codechicken/multipart/scalatraits/TTickableTile.scala
package codechicken.multipart.scalatraits

import java.util.{LinkedList => JLinkedList}

import codechicken.multipart.{TMultiPart, TileMultipart}
import net.minecraft.util.ITickable

import scala.collection.JavaConversions._

/**
 * Mixin Trait for parts implementing ITickable. Allows parts
 * to receive update calls every tick.
 */
trait TTickableTile extends TileMultipart with ITickable {

    private var doesTick = false
    var tickingParts = new JLinkedList[TMultiPart with ITickable]() //cache to reduce iteration of all parts

    override def copyFrom(that: TileMultipart) {
        super.copyFrom(that)
        that match {
            case tile: TTickableTile => {
                tickingParts = tile.tickingParts
                doesTick = tile.doesTick
            }
            case _ =>
        }
    }

    override def bindPart(part: TMultiPart) {
        super.bindPart(part)
        part match {
            case tickingPart: ITickable =>
                tickingParts += tickingPart
                setTicking(true)
            case _ =>
        }
    }

    override def partRemoved(part: TMultiPart, p: Int) {
        super.partRemoved(part, p)
        part match {
            case tickingPart: ITickable => tickingParts -= tickingPart
            //No need to disable ticking if empty, this tile will be replaced w/ nonticking one anyway
            case _ =>
        }
    }

    override def clearParts() {
        super.clearParts()
        tickingParts.clear()
    }

    override def update() {
        val it = tickingParts.iterator
        while (it.hasNext) {
            val p = it.next()
            if (p.tile != null) p.update()
        }
    }

    override def loadFrom(that: TileMultipart) {
        super.loadFrom(that)
        //Add the new tile to the ticking list.
        if (doesTick) {
            getWorld.tickableTileEntities.add(this)
        }
    }

    override def loadTo(that: TileMultipart) {
        super.loadTo(that)
        //Remove the old tile from the ticking list.
        if (doesTick) {
            getWorld.tickableTileEntities.remove(this)
        }
    }

    private def setTicking(tick: Boolean) {
        if (doesTick == tick) return
        doesTick = tick

        if (getWorld != null && getWorld.getTileEntity(getPos) == this) {
            if (tick) {
                getWorld.addTileEntity(this)
            } else {
                getWorld.tickableTileEntities.remove(this)
            }
        }
    }
}

/**
 * To handle obfuscation issues, this is registered as a java trait.
 */
class JTickableTile extends TileMultipart with TTickableTile {
}

File: src/main/java/codechicken/multipart/trait/TTickableTile.java
package codechicken.multipart.trait;

import codechicken.multipart.api.part.TickablePart;
import codechicken.multipart.api.part.MultiPart;
import codechicken.multipart.block.TileMultipart;
import codechicken.multipart.api.TickableTile;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.block.state.BlockState;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by covers1624 on 18/9/20.
 */
public class TTickableTile extends TileMultipart implements TickableTile {

    private final List<TickablePart> tickingParts = new ArrayList<>();

    public TTickableTile(BlockPos pos, BlockState state) {
        super(pos, state);
    }

    @Override
    public void copyFrom(TileMultipart that) {
        super.copyFrom(that);
        if (that instanceof TTickableTile) {
            tickingParts.clear();
            tickingParts.addAll(((TTickableTile) that).tickingParts);
        }
    }

    @Override
    public void bindPart(MultiPart part) {
        super.bindPart(part);
        if (part instanceof TickablePart p) {
            tickingParts.add(p);
        }
    }

    @Override
    public void partRemoved(MultiPart part, int p) {
        super.partRemoved(part, p);
        if (part instanceof TickablePart) {
            tickingParts.remove(part);
        }
    }

    @Override
    public void clearParts() {
        super.clearParts();
        tickingParts.clear();
    }

    @Override
    public void tick() {
        getCapCache().tick();
        for (int i = 0, tickingPartsSize = tickingParts.size(); i < tickingPartsSize; i++) {
            TickablePart part = tickingParts.get(i);
            if (((MultiPart) part).tile() != null) {
                part.tick();
            }
        }
    }
}
